File: ios/Modals/StepModal.swift
//
//  StepModal.swift
//  Glimpse
//
//  Created by Nitin Issac Joy on 10/07/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import RealmSwift

@objcMembers
class Image: Object {
  dynamic var image: String = "";
}

@objcMembers
class StepModal : Object {
  dynamic var stepId: Int = 0
  dynamic var tripId: String = ""
  dynamic var profileId: String = ""
  dynamic var stepName: String = ""
  dynamic var meanLatitude: Float64 = 0
  dynamic var meanLongitude: Float64 = 0
  dynamic var startTimestamp: TimeInterval = 0
  dynamic var endTimestamp: TimeInterval = 0
  dynamic var images: List<Image> = List<Image>()
  dynamic var markers: List<Region> = List<Region>()
  dynamic var masterImage: String = ""
  dynamic var distanceTravelled: Int = 0
  dynamic var desc: String = ""
  dynamic var temperature: String = ""
  
  func GetAsDictionary() -> [String:Any] {
    var dict: [String:Any] = [:]
    
    dict["stepId"] = self.stepId;
    dict["tripId"] = self.tripId;
    dict["profileId"] = self.profileId;
    dict["stepName"] = self.stepName;
    dict["meanLatitude"] = self.meanLatitude;
    dict["meanLongitude"] = self.meanLongitude;
    dict["startTimestamp"] = self.startTimestamp;
    dict["endTimestamp"] = self.endTimestamp;
    dict["masterImage"] = self.masterImage;
    dict["masterMarker"] = self.markers[0].GetAsDictionary();
    dict["distanceTravelled"] = self.distanceTravelled;
    dict["desc"] = self.desc;
    dict["temperature"] = self.temperature;
    dict["numberOfPicturesTaken"] = self.images.count;
    
    return dict;
  }
  
  func CopyConstructor(step: StepModal) {
    self.stepName = step.stepName;
    self.meanLatitude = step.meanLatitude;
    self.meanLongitude = step.meanLongitude;
    self.startTimestamp = step.startTimestamp;
    self.endTimestamp = step.endTimestamp;
    self.masterImage = step.masterImage;
    self.distanceTravelled = step.distanceTravelled;
    self.desc = step.desc;
    self.temperature = step.temperature;
  }
}

File: android/app/src/main/java/com/beerwithai/glimpse/Engine/Modals/StepModal.java
package com.beerwithai.glimpse.Engine.Modals;


import com.beerwithai.glimpse.Engine.Utils.ClusterProcessor;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.realm.RealmList;
import io.realm.RealmModel;
import io.realm.RealmObject;

public class StepModal implements RealmModel {
    public String profileId;
    public String tripId;
    public Integer stepId;
    public String stepName;
    public Float meanLatitude;
    public Float meanLongitude;
    public Double startTimestamp;
    public Double endTimestamp;
    public RealmList<String> images;
    public RealmList<Region> markers;
    public String masterImage;
    public Integer distanceTravelled;
    public String desc;
    public String temperature;

    public void CopyConstructor(StepModal step) {
        this.stepName = step.stepName;
        this.meanLatitude = step.meanLatitude;
        this.meanLongitude = step.meanLongitude;
        this.startTimestamp = step.startTimestamp;
        this.endTimestamp = step.endTimestamp;
        this.images = step.images;
        this.markers = step.markers;
        this.masterImage = step.masterImage;
        this.distanceTravelled = step.distanceTravelled;
        this.desc = step.desc;
        this.temperature  = step.temperature;
    }

    public Integer GetDistanceBetween(StepModal a) {

        ClusterModal _p = new ClusterModal();
        _p.latitude = a.meanLatitude;
        _p.longitude = a.meanLongitude;

        ClusterModal _q = new ClusterModal();
        _q.latitude = this.meanLatitude;
        _q.longitude = this.meanLongitude;

        return (ClusterProcessor.EarthDistance(_p, _q));
    }

    public WritableMap GetAsDictionary() {
        WritableMap obj = new WritableNativeMap();
        obj.putString("profileId", this.profileId);
        obj.putString("tripId", this.tripId);
        obj.putInt("stepId", this.stepId);
        obj.putString("stepName", this.stepName);
        obj.putDouble("meanLatitude", this.meanLatitude);
        obj.putDouble("meanLongitude", this.meanLongitude);
        obj.putDouble("startTimestamp", this.startTimestamp);
        obj.putDouble("endTimestamp", this.endTimestamp);
        obj.putString("masterImage", this.masterImage);
        obj.putMap("masterMarker", this.markers.get(0).GetAsDictionary());
        obj.putInt("distanceTravelled", this.distanceTravelled);
        obj.putString("desc", this.desc);
        obj.putString("temperature", this.temperature);
        obj.putInt("numberOfPicturesTaken", this.images.size());
        return obj;
    }
}